addons/mail/models/mail_activity.py

```diff
@@ -355,16 +357,25 @@ def write(self, values):
                         self.env['bus.bus'].sendone(
                             (self._cr.dbname, 'res.partner', partner.id),
                             {'type': 'activity_updated', 'activity_deleted': True})
+        if values.get('activity_type_id'):
+            for activity in self.filtered(lambda x: x.res_id and x.res_model):
+                self.env[activity.res_model].browse(activity.res_id)._set_activity_exception_type()
         return res
 
     @api.multi
     def unlink(self):
+        activitiy_records = []
         for activity in self:
             if activity.date_deadline <= fields.Date.today():
                 self.env['bus.bus'].sendone(
                     (self._cr.dbname, 'res.partner', activity.user_id.partner_id.id),
                     {'type': 'activity_updated', 'activity_deleted': True})
-        return super(MailActivity, self).unlink()
+            if activity.res_model and activity.res_id:
```

Comment:
and activity.activity_type_id.decoration_type in ['warning', 'danger']:

unlink is the reason why we have broken query count, i'd like to try to improve this a little bit.
If we remove an activity of another type, the decoration_type won't change. If decoration_type was warning or danger, it was because of another activity. If decoration_type was none, we don't care. Am I right?

Can you try this and check the new query count?